src/DataProtection/DataProtection/test/Microsoft.AspNetCore.DataProtection.Tests/KeyManagement/KeyRingProviderTests.cs

```diff
@@ -816,21 +816,24 @@ public async Task MultipleThreadsForceRefresh(bool failsToReadKeyRing)
         var expectedException = new InvalidOperationException(nameof(MultipleThreadsForceRefresh));
 
         var mockCacheableKeyRingProvider = new Mock<ICacheableKeyRingProvider>();
-        if (failsToReadKeyRing)
-        {
-            mockCacheableKeyRingProvider
-                .Setup(o => o.GetCacheableKeyRing(now))
-                .Throws(expectedException);
-        }
-        else
-        {
-            mockCacheableKeyRingProvider
-                .Setup(o => o.GetCacheableKeyRing(now))
-                .Returns(new CacheableKeyRing(
+        mockCacheableKeyRingProvider
+            .Setup(o => o.GetCacheableKeyRing(now))
+            .Returns<DateTimeOffset>(_ =>
+            {
+                // Simulate doing actual work.  We need this so that other threads have an opportunity
+                // to bypass the critical section.
+                Thread.Sleep(200);
```

Comment:
If this is a heuristic attempting to trigger a bug, might it be helpful to make this a random sleep (maybe averaging to 200ms)? Feel free to ignore.

Reply:
This callback will run within a critical section.  The goal is to have other threads try (and fail) to enter the critical section while it's held by this thread.  Increasing the amount of time in the critical section should increase the probability of that happening.  But I don't think randomness would improve our chances.

Reply:
Oops, this explanation was based on the old code.  The new explanation is substantially similar, except that there's basically shared data instead of a critical section.